lib/puppet/pops/types/types.rb

```diff
@@ -666,12 +677,21 @@ def _assignable?(o, guard)
 # @api public
 #
 class PScalarType < PAnyType
+  KNOWN_SCALARS = [String, Integer, Float, TrueClass, FalseClass, Regexp, Time::Timestamp, Time::Timespan, SemanticPuppet::Version, SemanticPuppet::VersionRange]
+  KNOWN_NON_SCALARS = [Array, Hash, PAnyType, NilClass]
+
   def self.register_ptype(loader, ir)
     create_ptype(loader, ir, 'AnyType')
   end
 
   def instance?(o, guard = nil)
-    assignable?(TypeCalculator.infer(o), guard)
+    if KNOWN_SCALARS.any? { |scalar_class| o.is_a?(scalar_class) }
+      true
+    elsif KNOWN_NON_SCALARS.any? { |scalar_class| o.is_a?(scalar_class) }
+      false
```

Comment:
I wrote [a performance test](/thallgren/ruby-performance/blob/master/case_versus_array_isa_match.rb) for this. The results surprised me a bit. The absolute fastest way to do this test is none of the ones we've been discussing. It's an if/elsif/else statement. I'll change the test accordingly.